File: src/main/java/io/github/saneea/dvh/feature/text/ConvertTextCase.java
package io.github.saneea.dvh.feature.text;

import java.io.Reader;
import java.io.Writer;

import org.apache.commons.cli.CommandLine;

import io.github.saneea.dvh.Feature;
import io.github.saneea.dvh.FeatureContext;

public abstract class ConvertTextCase implements Feature, Feature.CLI, Feature.In.Text.Reader, Feature.Out.Text.Writer {

	public static class Upper extends ConvertTextCase {
		public Upper() {
			super(Character::toUpperCase, Character::toUpperCase);
		}

		@Override
		public Meta meta(FeatureContext context) {
			return Meta.from("convert text to upper case (aBcDe -> ABCDE)");
		}
	}

	public static class Lower extends ConvertTextCase {
		public Lower() {
			super(Character::toLowerCase, Character::toLowerCase);
		}

		@Override
		public Meta meta(FeatureContext context) {
			return Meta.from("convert text to lower case (aBcDe -> abcde)");
		}
	}

	private final CharConverter charConverter;
	private final IntConverter intConverter;

	private Reader in;
	private Writer out;
	private CommandLine commandLine;

	public ConvertTextCase(CharConverter charConverter, IntConverter intConverter) {
		this.charConverter = charConverter;
		this.intConverter = intConverter;
	}

	@Override
	public void run(FeatureContext context) throws Exception {
		if (commandLine.hasOption(CommonOptions.NON_BUFFERED_STREAMS)) {
			int codePoint;
			while ((codePoint = in.read()) != -1) {
				out.write(intConverter.convert(codePoint));
				out.flush();
			}
		} else {
			char[] buf = new char[4096];
			int wasRead;
			while ((wasRead = in.read(buf)) != -1) {
				convertChars(buf, wasRead);
				out.write(buf, 0, wasRead);
			}
		}
	}

	private void convertChars(char[] buf, int size) {
		for (int i = 0; i < size; ++i) {
			buf[i] = convertChar(buf[i]);
		}
	}

	private char convertChar(char c) {
		return charConverter.convert(c);
	}

	@Override
	public void setIn(Reader in) {
		this.in = in;
	}

	@Override
	public void setOut(Writer out) {
		this.out = out;
	}

	@Override
	public void setCommandLine(CommandLine commandLine) {
		this.commandLine = commandLine;
	}

	@FunctionalInterface
	private interface CharConverter {
		char convert(char c);
	}

	@FunctionalInterface
	private interface IntConverter {
		int convert(int i);
	}

}

File: src/main/java/io/github/saneea/dvh/feature/text/ConvertTextCase.kt
package io.github.saneea.dvh.feature.text

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.CLI.CommonOptions
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import org.apache.commons.cli.CommandLine
import java.io.Reader
import java.io.Writer

class ConvertTextCase {

    class Upper :
        ConvertTextCaseBase(
            Character::toUpperCase,
            Character::toUpperCase
        ) {

        override val meta get() = Meta("convert text to upper case (aBcDe -> ABCDE)")
    }

    class Lower :
        ConvertTextCaseBase(
            Character::toLowerCase,
            Character::toLowerCase
        ) {

        override val meta get() = Meta("convert text to lower case (aBcDe -> abcde)")
    }
}

abstract class ConvertTextCaseBase(
    private val charConverter: (Char) -> Char,
    private val intConverter: (Int) -> Int
) :
    Feature,
    Feature.CLI,
    Feature.In.Text.Reader,
    Feature.Out.Text.Writer {

    override lateinit var context: FeatureContext
    override lateinit var inTextReader: Reader
    override lateinit var outTextWriter: Writer
    override lateinit var commandLine: CommandLine

    override fun run() {
        if (commandLine.hasOption(CommonOptions.NON_BUFFERED_STREAMS)) {
            var codePoint: Int
            while (inTextReader.read().also { codePoint = it } != -1) {
                outTextWriter.write(intConverter(codePoint))
                outTextWriter.flush()
            }
        } else {
            val buf = CharArray(4096)
            var wasRead: Int
            while (inTextReader.read(buf).also { wasRead = it } != -1) {
                convertChars(buf, wasRead)
                outTextWriter.write(buf, 0, wasRead)
            }
        }
    }

    private fun convertChars(buf: CharArray, size: Int) {
        for (i in 0 until size) {
            buf[i] = charConverter(buf[i])
        }
    }
}